staging/src/k8s.io/kube-apiextensions-server/pkg/controller/finalizer/crd_finalizer.go

```diff
@@ -0,0 +1,297 @@
+	client "k8s.io/kube-apiextensions-server/pkg/client/clientset/internalclientset/typed/apiextensions/internalversion"
+	informers "k8s.io/kube-apiextensions-server/pkg/client/informers/internalversion/apiextensions/internalversion"
+	listers "k8s.io/kube-apiextensions-server/pkg/client/listers/apiextensions/internalversion"
+)
+
+var cloner = conversion.NewCloner()
+
+// This controller finalizes the CRD by deleting all the CRs associated with it.
+type CRDFinalizer struct {
+	crdClient client.CustomResourceDefinitionsGetter
+	// clientPool is a dynamic client used to delete the individual instances
+	clientPool dynamic.ClientPool
+
+	crdLister listers.CustomResourceDefinitionLister
+	crdSynced cache.InformerSynced
+
+	// To allow injection for testing.
+	syncFn func(key string) error
+
+	queue workqueue.RateLimitingInterface
+}
+
+func NewCRDFinalizer(
+	crdInformer informers.CustomResourceDefinitionInformer,
+	crdClient client.CustomResourceDefinitionsGetter,
+	clientPool dynamic.ClientPool,
+) *CRDFinalizer {
+	c := &CRDFinalizer{
+		crdClient:  crdClient,
+		clientPool: clientPool,
+		crdLister:  crdInformer.Lister(),
+		crdSynced:  crdInformer.Informer().HasSynced,
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "CustomResourceDefinition-CRDFinalizer"),
+	}
+
+	crdInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addCustomResourceDefinition,
+		UpdateFunc: c.updateCustomResourceDefinition,
+	})
+
+	c.syncFn = c.sync
+
+	return c
+}
+
+func (c *CRDFinalizer) sync(key string) error {
+	cachedCRD, err := c.crdLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// no work to do
+	if cachedCRD.DeletionTimestamp.IsZero() || !hasFinalizer(cachedCRD) {
+		return nil
+	}
+
+	crd := &apiextensions.CustomResourceDefinition{}
+	if err := apiextensions.DeepCopy_apiextensions_CustomResourceDefinition(cachedCRD, crd, cloner); err != nil {
+		return err
+	}
+
+	// update the status condition.  This cleanup could take a while.
+	apiextensions.SetCRDCondition(crd, apiextensions.CustomResourceDefinitionCondition{
+		Type:    apiextensions.Terminating,
+		Status:  apiextensions.ConditionTrue,
+		Reason:  "InstanceDeletionInProgress",
+		Message: "CustomResource deletion is in progress",
+	})
+	crd, err = c.crdClient.CustomResourceDefinitions().UpdateStatus(crd)
+	if err != nil {
+		return err
+	}
+
+	// Its possible for a naming conflict to have removed this resource from the API after instances were created.
+	// For now we will cowardly stop finalizing.  If we don't go through the REST API, weird things may happen:
+	// no audit trail, no admission checks or side effects, finalization would probably still work but defaulting
+	// would be missed.  It would be a mess.
+	// This requires human intervention to solve, update status so they have a reason.
+	if !apiextensions.IsCRDConditionFalse(crd, apiextensions.NameConflict) {
```

Comment:
> Is this also going to trigger if you delete a CRD with name conflicts that was never installed (and so is safe to delete because it can't have CR instances)?

It will, but the detection of that isn't possible right now.  It is possible in general and useful to do in 1.7, but I'd rather not block this on that.  I will take it as a followup.

We can add a "was ever active" condition or some such to watch.